Bluetooth: Host: Avoid registering a callback twice

rugeGerritsen opened the pull request:
When a callback is registered twice, the slist becomes circular and we get stuck in a loop

In this PR we have choosen to ignore the second registration call if the callback has already been registed. The alternative is to trigger an assertion. That doesn't work if the assertions are turned off.

rugeGerritsen committed:
Bluetooth: Host: Avoid registering callback twice

Callbacks can only be registered once. Otherwise the slist
will become circular.

In this commit we have choosen to ignore the second registration
call if the callback has already been registed. The alternative
is to trigger an assertion. That doesn't work if the assertions
are turned off.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
Bluetooth: Host: Conn callback list to use slist

This allows us to use functionality provided by slist.
First use case: Avoid adding an element twice.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
Bluetooth: Host: Rename callback_list -> conn_cbs

This improves consistency with other callback lists like
scan_cbs and pa_sync_cbs.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1247,7 +1247,8 @@ struct bt_conn_cb {
 #endif /* CONFIG_BT_PATH_LOSS_MONITORING */
 
-	struct bt_conn_cb *_next;
+	/** Internally used field for list handling */
+	sys_snode_t node;
```
Should we name it 
```suggestion
	/** Internally used field for list handling */
	sys_snode_t _node;
```
To further indicate that this is internal? 

Does the doxygen tag `@internal` work for fields in structs?

rugeGerritsen replied:
I will give it a try

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -1588,7 +1588,11 @@ int bt_le_scan_stop(void)
 void bt_le_scan_cb_register(struct bt_le_scan_cb *cb)
 {
-	sys_slist_append(&scan_cbs, &cb->node);
+	sys_snode_t *prev_node;
+
+	if (!sys_slist_find(&scan_cbs, &cb->node, &prev_node)) {
+		sys_slist_append(&scan_cbs, &cb->node);
+	}
```
The documentation for the function doesn't seem to very explicit about it, but it looks like you can pass `NULL` to the prev_node
```suggestion
	if (!sys_slist_find(&scan_cbs, &cb->node, NULL)) {
		sys_slist_append(&scan_cbs, &cb->node);
	}
```

Thalley reviewed (commented):
Should the comment from https://github.com/zephyrproject-rtos/zephyr/pull/73894 be removed? 

Not sure if it makes sense to have it when it is now technically supported, even though it will just be a no-op.

alwa-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2444,40 +2467,22 @@ bt_security_t bt_conn_get_security(const struct bt_conn *conn)
-	callback_list = cb;
+	if (!sys_slist_find(&conn_cbs, &cb->_node, NULL)) {
+		sys_slist_append(&conn_cbs, &cb->_node);
+	}
```
What do you think about adding an error log here? I think double registration is indicative of problems in the application code.

OTOH, I can see the convenience for the application. But then if it's not an error, we should explain this in the API documentation.

```suggestion
	} else {
		/* Note to users: `bt_disable` does not unregister these. */
		LOG_ERR("Can't register bt_conn_cb, already in use");
	}
```

Thalley replied:
My opinion would be to add a return value to all register functions so that we can return `-EALREADY`, as we already do this with a significant subset of the register functions

rugeGerritsen replied:
Do you consider this as an API change or not? I can create a separate commit for it.
See https://github.com/zephyrproject-rtos/zephyr/pull/73894#issuecomment-2157705707

rugeGerritsen committed:
Bluetooth: Host: Callback registering functions to return status

Returning a status code will allow the application developer
to detect logic issues.

We consider this as not breaking the API.
If `-Werror -Wunused-result` is enabled, the application developer needs
to validate the return code.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1817,8 +1817,11 @@ int bt_le_per_adv_sync_delete(struct bt_le_per_adv_sync *per_adv_sync);
  * @param cb Callback struct. Must point to memory that remains valid.
+ *
+ * @retval 0 Success.
+ * @retval -ENOENT if @p cb was already registered.
```
Why `-ENOENT`? 

```
       ENOENT          No such file or directory (POSIX.1-2001).

                       Typically, this error results when a specified pathname does not exist, or one of the  com‐
                       ponents  in the directory prefix of a pathname does not exist, or the specified pathname is
                       a dangling symbolic link.
```

Not sure how that fits here to be honest. 
`-EALREADY`? `-EEXISTS`? 

I would also suggest to add tests of these new return values if possible (just as unit tests or something)

rugeGerritsen replied:
I can update that. Do you have a sample unit test that can be taken as a starting point? There seems like there are no unit tests for the host that can be used directly

alwa-nordic commented:
Tangentially relevant: I added an issue on thread-safely of `bt_conn_cb_register` and `bt_conn_cb_unregister`: https://github.com/zephyrproject-rtos/zephyr/issues/74173
